File: flutter_readium/android/src/main/kotlin/dk/nota/flutter_readium/models/ViewPortSize.kt
package dk.nota.flutter_readium.models

import dk.nota.flutter_readium.jsonDecode
import org.json.JSONObject
import org.readium.r2.shared.InternalReadiumApi

/**
 * Viewport information. Retrieved from javascript via window.flutterReadium.getViewPortSize().
 */
class ViewPortSize(
    /**
     * Viewport width
     */
    val width: Int,
    /**
     * Viewport height
     */
    val height: Int,
    /**
     * Vertical scroll offset
     */
    val scrollTop: Int,
    /**
     * Vertical scroll height, e.g. how much can we scroll vertically.
     */
    val scrollHeight: Int,
    /**
     * Horizontal scroll offset
     */
    val scrollLeft: Int,
    /**
     * Horizontal scroll width, e.g. how much can we scroll horizontally.
     */
    val scrollWidth: Int,
    /**
     * Scroll mode? If true, vertical scroll mode is enabled.
     */
    val scrollMode: Boolean,
) {
    /**
     * What is the previous progression, if we are scrolling backwards?
     */
    val prevProgression: Double
        get() =
            if (scrollMode) {
                (scrollTop.toDouble() - height.toDouble()) / scrollHeight.toDouble()
            } else {
                (
                    scrollLeft.toDouble() -
                        width.toDouble()
                ) /
                    scrollWidth.toDouble()
            }

    /**
     * What is the current progression? E.g top of the viewport?
     */
    val progression: Double
        get() = if (scrollMode) scrollTop.toDouble() / scrollHeight.toDouble() else scrollLeft.toDouble() / scrollWidth.toDouble()

    /**
     * What is the current bottom progression? E.g. the end of the viewport?
     *
     * This is needed to determine, if we need to scroll to the next file in the readingOrder.
     */
    val endProgression: Double
        get() =
            if (scrollMode) {
                (scrollTop.toDouble() + height.toDouble()) / scrollHeight.toDouble()
            } else {
                (
                    scrollLeft.toDouble() +
                        width.toDouble()
                ) /
                    scrollWidth.toDouble()
            }

    /**
     * Next progression, if we're scrolling forwards.
     *
     * If we reach >1.0 and [endProgression] is 1.0, we need to go to the next file in the readingOrder.
     */
    val nextProgression: Double
        get() =
            if (scrollMode) {
                (scrollTop.toDouble() + height.toDouble()) / scrollHeight.toDouble()
            } else {
                (
                    scrollLeft.toDouble() +
                        width.toDouble()
                ) /
                    scrollWidth.toDouble()
            }

    /**
     * Number of pages as calculated from the viewport size.
     */
    val numberOfPages: Double
        get() = if (scrollMode) scrollHeight.toDouble() / height.toDouble() else scrollWidth.toDouble() / width.toDouble()

    companion object {
        fun fromJson(
            json: String,
            scrollMode: Boolean,
        ): ViewPortSize = fromJson(jsonDecode(json) as JSONObject, scrollMode)

        @OptIn(InternalReadiumApi::class)
        fun fromJson(
            jsonObject: JSONObject,
            scrollMode: Boolean,
        ): ViewPortSize {
            val height = jsonObject.optInt("height")
            val width = jsonObject.optInt("width")
            val scrollHeight = jsonObject.optInt("scrollHeight")
            val scrollTop = jsonObject.optInt("scrollTop")
            val scrollWidth = jsonObject.optInt("scrollWidth")
            val scrollLeft = jsonObject.optInt("scrollLeft")

            return ViewPortSize(
                height = height,
                width = width,
                scrollTop = scrollTop,
                scrollHeight = scrollHeight,
                scrollLeft = scrollLeft,
                scrollWidth = scrollWidth,
                scrollMode = scrollMode,
            )
        }
    }
}

File: flutter_readium/ios/flutter_readium/Sources/flutter_readium/model/ViewPortSize.swift
import Foundation

public class ViewPortSize: Equatable {
  var width: Int
  var height: Int
  var scrollTop: Int
  var scrollHeight: Int
  var scrollLeft: Int
  var scrollWidth: Int
  var scrollMode: Bool

  init(
    width: Int,
    height: Int,
    scrollTop: Int,
    scrollHeight: Int,
    scrollLeft: Int,
    scrollWidth: Int,
    scrollMode: Bool
  ) {
    self.width = width
    self.height = height
    self.scrollTop = scrollTop
    self.scrollHeight = scrollHeight
    self.scrollLeft = scrollLeft
    self.scrollWidth = scrollWidth
    self.scrollMode = scrollMode
  }

  // MARK: - Computed properties (progressions)

  /// Previous progression when scrolling backwards
  var prevProgression: Double {
    if scrollMode {
      return Double(scrollTop - height) / Double(scrollHeight)
    } else {
      return Double(scrollLeft - width) / Double(scrollWidth)
    }
  }

  /// Current progression (top/left of viewport)
  var progression: Double {
    if scrollMode {
      return Double(scrollTop) / Double(scrollHeight)
    } else {
      return Double(scrollLeft) / Double(scrollWidth)
    }
  }

  /// End progression (bottom/right of viewport)
  var endProgression: Double {
    if scrollMode {
      return Double(scrollTop + height) / Double(scrollHeight)
    } else {
      return Double(scrollLeft + width) / Double(scrollWidth)
    }
  }

  /// Next progression when scrolling forwards
  var nextProgression: Double {
    // Same formula as endProgression in original Kotlin
    return endProgression
  }

  /// Number of pages based on viewport size
  var numberOfPages: Double {
    if scrollMode {
      return Double(scrollHeight) / Double(height)
    } else {
      return Double(scrollWidth) / Double(width)
    }
  }

  // MARK: - JSON

  static func fromJson(_ jsonString: String, scrollMode: Bool) -> ViewPortSize? {
    guard
      let data = jsonString.data(using: .utf8),
      let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    else { return nil }

    return fromJson(obj, scrollMode: scrollMode)
  }

  static func fromJson(_ json: [String: Any], scrollMode: Bool) -> ViewPortSize {
    let height = json["height"] as? Int ?? 0
    let width = json["width"] as? Int ?? 0
    let scrollHeight = json["scrollHeight"] as? Int ?? 0
    let scrollTop = json["scrollTop"] as? Int ?? 0
    let scrollWidth = json["scrollWidth"] as? Int ?? 0
    let scrollLeft = json["scrollLeft"] as? Int ?? 0

    return ViewPortSize(
      width: width,
      height: height,
      scrollTop: scrollTop,
      scrollHeight: scrollHeight,
      scrollLeft: scrollLeft,
      scrollWidth: scrollWidth,
      scrollMode: scrollMode
    )
  }

  func toJson() -> [String: Any] {
    return [
      "width": width,
      "height": height,
      "scrollTop": scrollTop,
      "scrollHeight": scrollHeight,
      "scrollLeft": scrollLeft,
      "scrollWidth": scrollWidth,
      "scrollMode": scrollMode
    ]
  }

  func toJsonString(pretty: Bool = false) -> String? {
    let options: JSONSerialization.WritingOptions = pretty ? [.prettyPrinted] : []
    guard let data = try? JSONSerialization.data(withJSONObject: toJson(), options: options)
    else { return nil }
    return String(data: data, encoding: .utf8)
  }

  // MARK: - Equatable

  public static func == (lhs: ViewPortSize, rhs: ViewPortSize) -> Bool {
    return lhs.width == rhs.width &&
      lhs.height == rhs.height &&
      lhs.scrollTop == rhs.scrollTop &&
      lhs.scrollHeight == rhs.scrollHeight &&
      lhs.scrollLeft == rhs.scrollLeft &&
      lhs.scrollWidth == rhs.scrollWidth &&
      lhs.scrollMode == rhs.scrollMode
  }
}
